File: jiuwenswarm/channels/ide/packages/vscode-extension/src/client/WsClient.ts
import WebSocket from 'ws';
import { JiuwenMessage } from './protocol';

export type WsStatus = 'disconnected' | 'connecting' | 'connected' | 'reconnecting';

export type WsEventMap = {
  status: (s: WsStatus) => void;
  message: (m: JiuwenMessage) => void;
  error: (e: Error) => void;
};

const BACKOFF_MS = [1000, 2000, 4000, 8000, 16000, 30000];
const DEFAULT_PING_INTERVAL_MS = 30_000;

export class WsClient {
  private ws: WebSocket | null = null;
  private status: WsStatus = 'disconnected';
  private retryCount = 0;
  private retryTimer: NodeJS.Timeout | null = null;
  private pingTimer: NodeJS.Timeout | null = null;
  private destroyed = false;
  private listeners: Partial<{ [K in keyof WsEventMap]: WsEventMap[K][] }> = {};
  // On macOS, Node.js 17+ resolves "localhost" to ::1 (IPv6) before 127.0.0.1 (IPv4).
  // If the server only listens on IPv4 the connection fails immediately. Force IPv4.
  private readonly url: string;

  constructor(
    url: string,
    private pingIntervalMs: number = DEFAULT_PING_INTERVAL_MS,
  ) {
    this.url = url.replace('://localhost:', '://127.0.0.1:');
  }

  on<K extends keyof WsEventMap>(event: K, listener: WsEventMap[K]): this {
    if (!this.listeners[event]) this.listeners[event] = [] as any;
    (this.listeners[event] as any[]).push(listener);
    return this;
  }

  off<K extends keyof WsEventMap>(event: K, listener: WsEventMap[K]): this {
    const arr = this.listeners[event] as any[] | undefined;
    if (arr) {
      const idx = arr.indexOf(listener);
      if (idx >= 0) arr.splice(idx, 1);
    }
    return this;
  }

  private emit<K extends keyof WsEventMap>(event: K, ...args: Parameters<WsEventMap[K]>) {
    const arr = this.listeners[event] as ((...a: any[]) => void)[] | undefined;
    arr?.forEach((fn) => fn(...args));
  }

  connect(): void {
    if (this.destroyed) return;
    this.setStatus('connecting');
    this.clearTimers();

    const ws = new WebSocket(this.url);
    this.ws = ws;

    ws.on('open', () => {
      if (this.destroyed) { ws.close(); return; }
      this.retryCount = 0;
      this.setStatus('connected');
      this.startPing();
    });

    ws.on('message', (data) => {
      try {
        const msg = JSON.parse(data.toString()) as JiuwenMessage;
        this.emit('message', msg);
      } catch {
        // ignore malformed frames
      }
    });

    ws.on('error', (err) => {
      this.emit('error', err);
    });

    ws.on('close', () => {
      this.stopPing();
      if (this.destroyed) return;
      this.ws = null;
      this.scheduleReconnect();
    });
  }

  /** Force a fresh reconnect (used for "New session"). */
  reconnect(): void {
    if (this.destroyed) return;
    this.clearTimers();
    this.retryCount = 0;
    if (this.ws) {
      try { this.ws.close(); } catch { /* ignore */ }
      this.ws = null;
    }
    this.connect();
  }

  send(msg: JiuwenMessage): boolean {
    if (this.ws?.readyState !== WebSocket.OPEN) return false;
    this.ws.send(JSON.stringify(msg));
    return true;
  }

  disconnect(): void {
    this.destroyed = true;
    this.clearTimers();
    this.stopPing();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.setStatus('disconnected');
  }

  getStatus(): WsStatus { return this.status; }
  isConnected(): boolean { return this.status === 'connected'; }

  private setStatus(s: WsStatus): void {
    if (this.status === s) return;
    this.status = s;
    this.emit('status', s);
  }

  private scheduleReconnect(): void {
    const delay = BACKOFF_MS[Math.min(this.retryCount, BACKOFF_MS.length - 1)];
    this.retryCount++;
    this.setStatus('reconnecting');
    this.retryTimer = setTimeout(() => {
      if (!this.destroyed) this.connect();
    }, delay);
  }

  setPingInterval(ms: number): void {
    this.pingIntervalMs = Math.max(5_000, Math.min(ms, 300_000));
    if (this.status === 'connected') {
      this.stopPing();
      this.startPing();
    }
  }

  private startPing(): void {
    if (this.pingIntervalMs <= 0) return;
    this.pingTimer = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.ping();
      }
    }, this.pingIntervalMs);
  }

  private stopPing(): void {
    if (this.pingTimer) { clearInterval(this.pingTimer); this.pingTimer = null; }
  }

  private clearTimers(): void {
    if (this.retryTimer) { clearTimeout(this.retryTimer); this.retryTimer = null; }
  }
}

File: jiuwenswarm/channels/ide/packages/jetbrains-plugin/src/main/kotlin/com/jiuwenswarm/plugin/client/WsClient.kt
package com.jiuwenswarm.plugin.client

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.logger
import okhttp3.*
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

private val LOG = logger<WsClient>()

enum class WsStatus { DISCONNECTED, CONNECTING, CONNECTED, RECONNECTING }

typealias StatusListener = (WsStatus) -> Unit
typealias MessageListener = (JsonObject) -> Unit

private val BACKOFF_SECONDS = longArrayOf(1, 2, 4, 8, 16, 30)
private val gson = Gson()

class WsClient(
    url: String,
    pingIntervalSeconds: Long = 15,
) : Disposable {

    // On macOS, "localhost" resolves to ::1 (IPv6) before 127.0.0.1 (IPv4).
    // If the server only listens on IPv4, OkHttp gets "Connection Refused" on ::1
    // and stays in reconnect loop. Force IPv4 to avoid this.
    private val url = url.replace("://localhost:", "://127.0.0.1:")

    @Volatile private var pingIntervalSec: Long = pingIntervalSeconds.coerceIn(0, 300)

    private var client: OkHttpClient = buildClient()

    private fun buildClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .readTimeout(0, TimeUnit.MILLISECONDS)
        if (pingIntervalSec > 0) {
            builder.pingInterval(pingIntervalSec, TimeUnit.SECONDS)
        }
        return builder.build()
    }

    /** Rebuild the underlying OkHttp client with a new keep-alive interval. */
    fun setPingInterval(seconds: Long) {
        val clamped = seconds.coerceIn(0, 300)
        if (pingIntervalSec == clamped) return
        pingIntervalSec = clamped
        client = buildClient()
    }

    private val scheduler = Executors.newSingleThreadScheduledExecutor { r ->
        Thread(r, "jiuwenswarm-ws-scheduler").also { it.isDaemon = true }
    }

    @Volatile private var ws: WebSocket? = null
    @Volatile private var status = WsStatus.DISCONNECTED
    private val retryCount = AtomicInteger(0)
    private var retryFuture: ScheduledFuture<*>? = null
    @Volatile private var destroyed = false

    private val statusListeners = CopyOnWriteArrayList<StatusListener>()
    private val messageListeners = CopyOnWriteArrayList<MessageListener>()

    fun addStatusListener(l: StatusListener) = statusListeners.add(l)
    fun removeStatusListener(l: StatusListener) = statusListeners.remove(l)
    fun addMessageListener(l: MessageListener) = messageListeners.add(l)
    fun removeMessageListener(l: MessageListener) = messageListeners.remove(l)

    fun connect() {
        if (destroyed) return
        retryFuture?.cancel(false)
        setStatus(WsStatus.CONNECTING)
        val request = Request.Builder().url(url).build()
        ws = client.newWebSocket(request, Listener())
    }

    /** Force a fresh reconnect (used for "New session"). */
    fun reconnect() {
        if (destroyed) return
        retryFuture?.cancel(false)
        retryCount.set(0)
        ws?.close(1000, "User requested reconnect")
        ws = null
        connect()
    }

    fun send(json: JsonObject): Boolean {
        val sock = ws ?: return false
        return sock.send(gson.toJson(json))
    }

    fun isConnected() = status == WsStatus.CONNECTED

    fun getStatus() = status

    override fun dispose() {
        destroyed = true
        retryFuture?.cancel(false)
        ws?.close(1000, "Plugin disposed")
        ws = null
        scheduler.shutdownNow()
        client.dispatcher.executorService.shutdown()
        setStatus(WsStatus.DISCONNECTED)
    }

    private fun setStatus(s: WsStatus) {
        if (status == s) return
        status = s
        statusListeners.forEach { it(s) }
    }

    private fun scheduleReconnect() {
        if (destroyed) return
        val delay = BACKOFF_SECONDS[minOf(retryCount.getAndIncrement(), BACKOFF_SECONDS.size - 1)]
        setStatus(WsStatus.RECONNECTING)
        LOG.info("Reconnecting in ${delay}s (attempt ${retryCount.get()})")
        retryFuture = scheduler.schedule({ if (!destroyed) connect() }, delay, TimeUnit.SECONDS)
    }

    private inner class Listener : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            LOG.info("Connected to JiuwenSwarm at $url")
            retryCount.set(0)
            setStatus(WsStatus.CONNECTED)
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            try {
                val obj = gson.fromJson(text, JsonObject::class.java)
                messageListeners.forEach { it(obj) }
            } catch (e: Exception) {
                LOG.warn("Failed to parse message: ${text.take(100)}", e)
            }
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            LOG.warn("WebSocket error", t)
            ws = null
            scheduleReconnect()
        }

        override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
            webSocket.close(1000, null)
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            ws = null
            if (!destroyed) scheduleReconnect()
        }
    }
}
